crates/key-value/src/host_component.rs

```diff
@@ -66,4 +67,27 @@ impl DynamicHostComponent for KeyValueComponent {
 
         Ok(())
     }
+
+    fn validate_app(&self, app: &spin_app::App) -> anyhow::Result<()> {
+        let mut errors = vec![];
+
+        for component in app.components() {
+            let store_manager = self.manager.get(&component);
+            for allowed in component
+                .get_metadata(KEY_VALUE_STORES_KEY)?
+                .unwrap_or_default()
+            {
+                if !store_manager.is_defined(&allowed) {
+                    let err = format!("Component {} is granted access to key-value store '{allowed}', which is not defined", component.id());
```

Comment:
I'm not sure that we need to frame `key_value_stores` in terms of access control. Maybe something like:

```suggestion
                    let err = format!("Component {} uses the key-value store '{allowed}', which is not defined", component.id());
```